src/core/lib/gprpp/time.h

```diff
@@ -49,8 +49,9 @@
     uint64_t now = grpc_core::Timestamp::FromTimespecRoundDown( \
                        gpr_now(GPR_CLOCK_MONOTONIC))            \
                        .milliseconds_after_process_epoch();     \
-    uint64_t prev_tsamp = prev.exchange(now);                   \
-    if (now - prev_tsamp > (n) * 1000) {                        \
+    if (prev == 0) prev = now;                                  \
```

Comment:
this assignment is racey right?

Reply:
In terms of memory access errors? I don't see how, this macro should be called in a loop, single-threaded, in which these would be local variables. In terms of comparison in the subsequent equality check, it's ok if there's some time drift here, this log rate can be loose.

See `GRPC_LOG_EVERY_N_SEC` above. This change is a fix, it was inconsistently logging at the expected rate.

Reply:
I was thinking just about logical raciness affecting the time/rate.

e..g. two threads could write to `prev` on lines 52 and 54.

Though as discussed, neither of these cases shouldn't really make a meaningful difference in the rate math

Reply:
I see your point, good catch. As discussed, there is some potential time slop here, we could improve it using something like a thread_local, but I agree it's likely not worth it. This method may already log messages slower than `n` if it's called infrequently, or called with some rate that does not evenly divide `n`.